vulnerabilities/msgvalue-loop.md

```diff
@@ -0,0 +1,36 @@
+# Using ``msg.value`` in a loop
+
+The particularity with msg.value is that ``msg.value`` is set on every contract call and not on the transaction level. Thus, if the function with loop is transferring ``msg.value`` on every iteration, all the ``msg.value`` sent to the function is transferred in the first iteration and other iteration will proceed with ``msg.value = 0`` which may cause unexpected consequences.
```

Comment:
I think the last sentence here is a bit confusing, it seems to suggest that msg.value will be set as 0 but won't it stay the same?

Reply:
The first iteration transfers all the ``msg.value`` sent to the function to the recipient while looping. It won't stay the same as during ``msg.value transfer``, contract call happens and ``msg.value`` is reset.

``msg.value`` value only stays the same if it is not transferred and the second example deals exactly with that.

Reply:
I don't think this is right, because e.g. in the `depositer` example, msg.value is the value provided with the call to `depositer.deposit` and it should attempt to use the same value each time iiuc. Is that not right?

Reply:
Yeah you are right. 
```solidity
contract depositer {
    function deposit(address weth) payable external {
        for (uint i = 0; i < 5; i ++) {
            WETH(weth).deposit{value: msg.value}(); 
        }
    }
}
```
Let's say ``msg.value`` sent is ``1e18``. 

That ``1e18`` will be deposited in the first iteration here: ``WETH(weth).deposit{value: msg.value}();``
Now in the second iteration,  Again, the function attempts to deposit ``1e18`` but ``1e18`` has already been used in the first iteration. Thus ``0`` value will be deposited from the second iteration.

Reply:
It's also worth noting that if the contract holds more ETH then the transactions could actually succeed, allowing an attacker to unexpectedly spend the contract's balance

Reply:
That's an interesting idea. Will the function use contracts ETH from ``msg.value``  when msg.value determines the amount sent to the payable function. I will test this in Remix.

Reply:
Fixed in the latest commit!